File: lib/client/ts/clientledger.ts
import * as flatbuffers from "flatbuffers";
import axios from "axios";
import { ECSComponent } from "./ecs";
import { CommitDiffT, CommitProposal, CommitProposalT, ComponentIdentifierT, ComponentT, uuidv4T } from "../../schema/bimrepo";
import { CommitResponse, CommitResponseT } from "../../schema/bimrepo/commit-response";

export class IServerLedger
{
    private address: string;
    protected notifyHeadChanged: (head: number) => void | undefined;
    private ws: WebSocket;
    
    constructor(address: string)
    {
        this.address = address;

        if (this.address != "")
        {
            this.ws = new WebSocket(
                `${this.address}/ws`
            );
            this.ws.onmessage = (event) => {
                let head = event.data;
                if (this.notifyHeadChanged)
                {
                    this.notifyHeadChanged(head);
                }
            };
        }
    }

    public async Commit(commitBuffer: Uint8Array)
    {
        let response = await axios.post(`${this.address}/commit`, 
            commitBuffer.buffer
        , {
            headers: {'Content-Type': 'application/octet-stream'},
        });

        return toArrayBuffer(response.data);
    }

    public async GetCommit(id: number)
    {
        let response = await axios.get(`${this.address}/commit/${id}`, {
            responseType: "arraybuffer"
        });
        return toArrayBuffer(response.data);
    }

    public async NotifyHeadChanged(headChanged: (head: number)=>void)
    {
        this.notifyHeadChanged = headChanged;
    }
}

function toArrayBuffer(buffer) {
    const arrayBuffer = new ArrayBuffer(buffer.length);
    const view = new Uint8Array(arrayBuffer);
    for (let i = 0; i < buffer.length; ++i) {
      view[i] = buffer[i];
    }
    return view;
  }

export default class ClientLedger {

    private modifiedComponents: ECSComponent[] = [];
    private serverLedger: IServerLedger;
    private notifyHeadChanged: (head: number) => void | undefined;
    private observedCommits: number[] = [];

    constructor(serverLedger: IServerLedger, )
    {
        this.serverLedger = serverLedger;
        serverLedger.NotifyHeadChanged(this.NotifyHeadChanged.bind(this));
    }

    public SetNotifyHeadChanged(headChanged: (head: number) => void)
    {
        this.notifyHeadChanged = headChanged;
    }

    private NotifyHeadChanged(head: number)
    {
        this.observedCommits.push(head);
        if (this.notifyHeadChanged)
        {
            this.notifyHeadChanged(head);
        }
    }

    update(component: ECSComponent)
    {
        // TODO: check for duplicates
        this.modifiedComponents.push(component);
    }

    canCommit()
    {
        return this.modifiedComponents.length !== 0;
    }

    private ComponentToIdentifier(component: ECSComponent)
    {
        let id = new ComponentIdentifierT();

        // entity id
        let uuidv4 = new uuidv4T();
        uuidv4.values = [...component.getEntityID().bytes.values()];
        id.entity = uuidv4;

        id.typeHash = component.getTypeHash();
        id.componentIndex = 0; // temporarily hardcoded 1 component per entity

        return id;
    }

    async commit(author: string, message: string): Promise<number>
    {
        if (!this.canCommit()) throw new Error(`No components to commit`);

        let commit = new CommitProposalT(); 

        commit.author = author;
        commit.message = message;

        commit.diff = new CommitDiffT();

        let exportedTypes = {};

        this.modifiedComponents.forEach((component) => {
            let id = this.ComponentToIdentifier(component);
            let exported = component.exportToDataArray(id);
            commit.diff?.updatedComponents.push(exported);
            let name = component.getSimplifiedName();
            if (!exportedTypes[name])
            {
                exportedTypes[name] = true;
                let data = component.exportDefinitionToArray();
                commit.diff?.updatedSchemas.push(data);
            }
        });

        let fbb = new flatbuffers.Builder(1);

        CommitProposal.finishCommitProposalBuffer(fbb, commit.pack(fbb));

        let requestBuffer = fbb.asUint8Array().slice(0);

        let responseBuffer = await this.serverLedger.Commit(requestBuffer);

        let commitResponse = new CommitResponseT();
        CommitResponse.getRootAsCommitResponse(new flatbuffers.ByteBuffer(responseBuffer)).unpackTo(commitResponse);

        return commitResponse.id;
    }

    async GetCommit(id: number)
    {
        let responseBuffer = await this.serverLedger.GetCommit(id);

        let commit = new CommitProposalT();
        CommitProposal.getRootAsCommitProposal(new flatbuffers.ByteBuffer(responseBuffer)).unpackTo(commit);

        return commit;
    }
}

File: lib/client/cs/clientledger.cs

using bimrepo;
using Google.FlatBuffers;
using System.Net.WebSockets;
using System;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.IO;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Net.Http;


public class ClientLedger {

    private string address;
    private string ledgername;
    private List<ECSComponent> modifiedComponents = new List<ECSComponent>();

    public ClientLedger(string address, string ledgername)
    {
        this.address = address;
        this.ledgername = ledgername;
      }

    public async Task Listen(Action<int> headChanged)
    {
        var client = new ClientWebSocket();
        await client.ConnectAsync(new Uri($"ws://{this.address}/ws"), CancellationToken.None);

        var ReceiveBufferSize = 1024;
        var buffer = new byte[ReceiveBufferSize];
        while (true)
        {
            var receiveResult = await client.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);

            string s = Encoding.UTF8.GetString(buffer, 0, receiveResult.Count);

            int commitId = int.Parse(s);    

            headChanged(commitId);
            
            if (receiveResult.MessageType == WebSocketMessageType.Close)
            {
                break;
            }
        }
    }

    public static byte[] ReadFully(Stream input)
    {
        byte[] buffer = new byte[16 * 1024];
        using (MemoryStream ms = new MemoryStream())
        {
            int read;
            while ((read = input.Read(buffer, 0, buffer.Length)) > 0)
            {
                ms.Write(buffer, 0, read);
            }
            return ms.ToArray();
        }
    }

    public void update(ECSComponent component)
    {
        // TODO: check for duplicates
        this.modifiedComponents.Add(component);
    }

    public bool canCommit()
    {
        return this.modifiedComponents.Count != 0;
    }

    private ComponentIdentifierT ComponentToIdentifier(ECSComponent component)
    {
        var id = new ComponentIdentifierT();

        // entity id
        var uuidv4 = new uuidv4T();
        uuidv4.Values = component.getEntityID().bytes;
        id.Entity = uuidv4;

        id.TypeHash = component.getTypeHash();
        id.ComponentIndex = 0; // temporarily hardcoded 1 component per entity

        return id;
    }

    private string GetURL()
    {
        return $"http://{this.address}/data/{this.ledgername}";
    }

    public async Task<int> commit(string author, string message)
    {
        if (!this.canCommit()) throw new Exception($"No components to commit");

        var commit = new CommitProposalT(); 

        commit.Author = author;
        commit.Message = message;

        commit.Diff = new CommitDiffT();
        commit.Diff.UpdatedSchemas = new List<SchemaT>();
        commit.Diff.UpdatedComponents = new List<ComponentT>();

        var exportedTypes = new Dictionary<string, bool>();

        this.modifiedComponents.ForEach((component) => {
            var id = this.ComponentToIdentifier(component);
            var exported = component.exportToDataArray(id);
            commit.Diff?.UpdatedComponents.Add(exported);
            var name = component.getSimplifiedName();
            if (!exportedTypes.ContainsKey(name))
            {
                exportedTypes[name] = true;
                var data = component.exportDefinitionToArray();
                commit.Diff?.UpdatedSchemas.Add(data);
            }
        });

        var fbb = new FlatBufferBuilder(1);

        fbb.Finish(CommitProposal.Pack(fbb, commit).Value);

        var requestURL = $"{this.GetURL()}/commit";
        // do request with fbb
        var arr = fbb.DataBuffer.ToSizedArray();
        var content = new ByteArrayContent(arr);
        HttpClient client = new HttpClient();
        content.Headers.ContentType = new MediaTypeWithQualityHeaderValue("application/octet-stream");
        var response = await client.PostAsync(requestURL, content);
        var responseStream = await response.Content.ReadAsStreamAsync();

        var bytes = ReadFully(responseStream);

        var commitResponse = CommitResponseT.DeserializeFromBinary(bytes);

        this.modifiedComponents.Clear();

        return commitResponse.Id;
    }

    public async Task<CommitProposalT> GetCommit(int id)
    {
        HttpClient client = new HttpClient();
        var url = $"{this.GetURL()}/commit/{id}";
        var response = await client.GetAsync(url);
        var responseStream = await response.Content.ReadAsStreamAsync();
        var bytes = ReadFully(responseStream);

        return CommitProposalT.DeserializeFromBinary(bytes);
    }
}